spinalcordtoolbox/scripts/sct_propseg.py

```diff
@@ -450,6 +455,11 @@ def propseg(img_input, options_dict):
     if not os.path.exists(folder_output):
         os.makedirs(folder_output)
 
+    if arguments.o is not None:
+        fname_out = arguments.o
+    else:
+        fname_out = os.path.basename(add_suffix(fname_data, "_seg"))
+
```

Comment:
This is the cause of the error in https://github.com/neuropoly/spinalcordtoolbox/pull/3081#issuecomment-752034463. It needs to be aware of `-ofolder` too.

**Probably** this PR should **remove** `-ofolder` entirely since `-o` covers the same purpose. You should replace the above `folder_out` section with:

```
folder_out = os.path.dirname(arguments.o)
```

If you need to keep both arguments -- which I really recommend against -- then you need to make sure they are consistent with each other. That means doing something like:

```
# quick draft, don't commit as is, please test it out
folder_out, fname_out = os.path.split(arguments.o)
if arguments.ofolder is not None and arguments.ofolder != folder_out:
    raise ValueError(f"Output file {arguments.o} must be in output folder {arguments.ofolder}") 
```

What's `-ofolder` for anyway? Does `propseg` output extra metadata to that folder?

Reply:
By the time it gets to

https://github.com/neuropoly/spinalcordtoolbox/blob/552b908c63234c23cf65e9e520ec3c88a55bcec8/spinalcordtoolbox/scripts/sct_propseg.py#L635-L640

you must have

```
assert not os.path.dirname(fname_out)
```

That's what that section of code is implicitly assuming. You can either fix it there, or fix it above. Adding that assertion is probably a good idea, at least for the tests.

Reply:
Ou.... j'ai pas encore tout lire. Il y ..trois? fichiers? `f"{F}_rescaled.nii.gz"` et `f"{F}_rescaled_centerline.nii.gz"` et ça veux redire vers `f"{F}.nii.gz"` et `f"{F}_centerline.nii.gz"?

Est-ce qu'on c'est possible de juste demander de `isct_propseg` de faire le bon chose en prèmiere?

Reply:
> What's -ofolder for anyway? Does propseg output extra metadata to that folder?

propseg can output more than one stuff, e.g. `-CSF`, ` -centerline-coord`. Historically, the `-ofolder` was created when we were using `sct_testing`: this script included a function that used the flag `-ofolder`, when it existed in some scripts. Since then, we completely refactored the testing suite and we don't use -ofolder anymore, so I agree we could get rid of -ofolder as it is confusing.

But we need to keep in mind the multiple (possible) outputs. However, since these are not widely used, we can keep -o to specify the output of the cord segmentation (primary interest for sct_propseg) and have the other outputs with fix naming (very few people use the other outputs anyway).

Reply:
I tried something to remove this -ofolder issue, tell me what you think! (bab7bff)

Reply:
If the -ofolder issue is fixed, `tmp_path` should be usable again, so could you add `tmp_path` back to the `sct_propseg` test?  Would help to test this change. :slightly_smiling_face:

Reply:
To clarify what @kousu was saying, the error occurs with this line:

https://github.com/neuropoly/spinalcordtoolbox/blob/fd00c800467aa194fcbae073eb10bc589d772920/spinalcordtoolbox/scripts/sct_propseg.py#L639-L640

When called in CI, the two values are:

Reply:
To clarify what @kousu was saying, the error occurs with this line:

https://github.com/neuropoly/spinalcordtoolbox/blob/fd00c800467aa194fcbae073eb10bc589d772920/spinalcordtoolbox/scripts/sct_propseg.py#L639-L640

When this command is called in the test, the two values are this: [(CI link)](https://travis-ci.com/github/neuropoly/spinalcordtoolbox/jobs/465807495#L2096-L2105)

```
src = './t2_rescaled_seg.nii.gz'
dst = '/tmp/pytest-of-travis/pytest-0/popen-gw0/test_sct_propseg_o_flag0/test_seg.nii.gz'
```

The problem here is that `src` is not using `tmp_path`. (Every output file should be created in the `tmp_path` directory.)

To fix this, you need to make sure that `folder_output` uses `tmp_path`.

https://github.com/neuropoly/spinalcordtoolbox/blob/3df9aeffcade01e30befd9b6074fc77663d051e5/spinalcordtoolbox/scripts/sct_propseg.py#L438

So, you might use [`os.path.dirname`](https://docs.python.org/3/library/os.path.html#os.path.dirname) get the parent directory from the `-o` argument, so that everything is output to that directory.

```diff
- folder_output = './' 
+ folder_output = os.path.dirname(fname_out)  # Get base directory from output path
```

Reply:
@joshuacwnewton what do you think about f81733d ?

Reply:
Looks good to me! :)